openedx/core/djangoapps/user_api/accounts/serializers.py

```diff
@@ -573,26 +576,47 @@ def validate_new_name(self, new_name):
             raise serializers.ValidationError('Name cannot contain a URL')
 
 
-def get_extended_profile(user_profile):
+def get_extended_profile(user_profile: UserProfile) -> list[dict[str, str]]:
     """
-    Returns the extended user profile fields stored in user_profile.meta
+    Retrieve extended user profile fields for API serialization.
+
+    This function extracts custom profile fields that extend beyond the standard
+    UserProfile model. It first attempts to get data from a custom extended profile
+    model (if configured), then falls back to the user_profile.meta JSON field.
+    The returned data is filtered to include only fields specified in the
+    'extended_profile_fields' site configuration.
+
+    The function supports two data sources:
+    1. Custom model: If `REGISTRATION_EXTENSION_FORM` setting points to a form with
+       a `Meta.model`, data is retrieved from that model using `model_to_dict()`
+    2. Fallback: JSON data stored in `UserProfile.meta` field
+
+    Args:
+        user_profile (UserProfile): The user profile instance to get extended fields from.
+
+    Returns:
+        list[dict[str, str]]: A list of dictionaries, each containing:
+            - 'field_name': The name of the extended profile field
+            - 'field_value': The value of the field (converted to string)
     """
+    def get_extended_profile_data():
+        extended_profile_model = get_extended_profile_model()
 
-    # pick the keys from the site configuration
-    extended_profile_field_names = configuration_helpers.get_value('extended_profile_fields', [])
+        if extended_profile_model:
+            try:
+                profile_obj = extended_profile_model.objects.get(user=user_profile.user)
+                return model_to_dict(profile_obj)
+            except (AttributeError, extended_profile_model.DoesNotExist):
+                return {}
 
-    try:
-        extended_profile_fields_data = json.loads(user_profile.meta)
-    except ValueError:
-        extended_profile_fields_data = {}
+        try:
+            return json.loads(user_profile.meta or "{}")
+        except (ValueError, TypeError, AttributeError):
+            return {}
```

Comment:
This section made me question something: is the profile meta and the extended profile model out of date at some point? Should they be in sync?